From bb70c97c0b14b9f7139a6ed1092ba124e3e1de73 Mon Sep 17 00:00:00 2001
From: Jiafei Pan <Jiafei.Pan@nxp.com>
Date: Fri, 16 Jan 2026 17:27:18 +0800
Subject: [PATCH 14/15] plat: imx: dvfs: bring all running Core into WFE

In general, the requriement for DRAM frequency changing is raised by Linux,
the current implementation is only bring all the CPU Core used by Linux to
be WFE state in order to DDR access in the bus frequency changing process.

But in a flexible AMP system, some Cortex-A Core maybe run RTOS or baremetal,
these CPU cores should also go into WFE state in the bus frequency changing
process.

This patch is to bring all running Cores into WFE.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
Upstream-Status: Pending
---
 plat/imx/common/imx_intercore_svc.c         |  8 ++---
 plat/imx/common/include/imx_intercore_svc.h |  4 +--
 plat/imx/imx8m/ddr/dram.c                   | 30 +++++++++++++---
 plat/imx/imx93/ddr/dram.c                   | 39 +++++++++++++++++----
 4 files changed, 62 insertions(+), 19 deletions(-)

diff --git a/plat/imx/common/imx_intercore_svc.c b/plat/imx/common/imx_intercore_svc.c
index fc855cca6..51d2f44a8 100644
--- a/plat/imx/common/imx_intercore_svc.c
+++ b/plat/imx/common/imx_intercore_svc.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2024 NXP
+ * Copyright 2024, 2026 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -42,17 +42,15 @@ int register_intercore_dvfs_svc(interrupt_type_handler_t handler)
 	return 0;
 }
 
-int trigger_intercore_dvfs_svc(void)
+int trigger_intercore_dvfs_svc(uint32_t running_cores)
 {
-	unsigned int cpu_id = get_cpu_id();
 	percpu_svc_inf_t *inf;
 	uint32_t i;
 
 	for (i = 0; i < PLATFORM_CORE_COUNT; i++) {
 		inf = &svc_info[i];
 		inf->dvfs_svc.trigged = true;
-		/* Skip raise SGI for current CPU */
-		if (i != cpu_id) {
+		if (running_cores & (0x1 << (i * 8))) {
 			plat_ic_raise_el3_sgi(INTERCORE_SVC_IRQ, logical_id_to_hwid(i));
 		}
 	}
diff --git a/plat/imx/common/include/imx_intercore_svc.h b/plat/imx/common/include/imx_intercore_svc.h
index b045bda29..2ba84d775 100644
--- a/plat/imx/common/include/imx_intercore_svc.h
+++ b/plat/imx/common/include/imx_intercore_svc.h
@@ -1,5 +1,5 @@
 /*
- * Copyright 2024 NXP
+ * Copyright 2024, 2026 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -35,7 +35,7 @@ static inline unsigned int get_cpu_id(void)
 }
 
 int register_intercore_dvfs_svc(interrupt_type_handler_t handler);
-int trigger_intercore_dvfs_svc(void);
+int trigger_intercore_dvfs_svc(uint32_t running_cores);
 int register_intercore_cpuoff_svc(interrupt_type_handler_t handler);
 int trigger_intercore_cpuoff_svc(unsigned int cpu);
 void imx_intercore_svc_init(void);
diff --git a/plat/imx/imx8m/ddr/dram.c b/plat/imx/imx8m/ddr/dram.c
index d1e66dbfe..7e2f94e18 100644
--- a/plat/imx/imx8m/ddr/dram.c
+++ b/plat/imx/imx8m/ddr/dram.c
@@ -1,10 +1,11 @@
 /*
- * Copyright 2019-2024 NXP
+ * Copyright 2019-2024, 2026 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
 
 #include <common/runtime_svc.h>
+#include <lib/el3_runtime/cpu_data.h>
 #include <lib/mmio.h>
 #include <lib/spinlock.h>
 
@@ -322,13 +323,19 @@ int dram_dvfs_get_freq_info(void *handle, u_register_t index)
 	}
 }
 
+static inline aff_info_state_t dvfs_get_aff_info_state_by_idx(unsigned int idx)
+{
+	return get_cpu_data_by_index(idx,
+				     psci_svc_cpu_data.aff_info_state);
+}
+
 int dram_dvfs_handler(uint32_t smc_fid, void *handle,
 	u_register_t x1, u_register_t x2, u_register_t x3)
 {
 	uint64_t mpidr = read_mpidr_el1();
 	unsigned int cpu_id = MPIDR_AFFLVL0_VAL(mpidr);
 	unsigned int fsp_index = x1;
-	uint32_t online_cores = x2;
+	uint32_t running_cores = 0;
 
 	if (x1 == IMX_SIP_DDR_DVFS_GET_FREQ_COUNT) {
 		SMC_RET1(handle, dram_info.num_fsp);
@@ -338,9 +345,23 @@ int dram_dvfs_handler(uint32_t smc_fid, void *handle,
 		wait_ddrc_hwffc_done = true;
 		dsb();
 
+		/* Check all running CPU Cores */
+		for (unsigned int i = 0; i < PLATFORM_CORE_COUNT; i++) {
+			/* Skip current CPU Core */
+			if (i == cpu_id) {
+				continue;
+			}
+			/* Perform cache maintanence ahead of reading the target CPU state */
+			flush_cpu_data_by_index(i, psci_svc_cpu_data.aff_info_state);
+			if (dvfs_get_aff_info_state_by_idx(i) == AFF_STATE_ON) {
+				running_cores |= (0x1 << (i * 8));
+			}
+		}
+
 		/* trigger the SGI IPI to info other cores */
-		trigger_intercore_dvfs_svc();
+		trigger_intercore_dvfs_svc(running_cores);
 #if defined(PLAT_imx8mq)
+		uint32_t online_cores = x2;
 		for (unsigned int i = 0; i < PLATFORM_CORE_COUNT; i++) {
 			if (i != cpu_id && online_cores & (1 << (i * 8))) {
 				imx_gpc_core_wake(1 << i);
@@ -348,9 +369,8 @@ int dram_dvfs_handler(uint32_t smc_fid, void *handle,
 		}
 #endif
 		/* make sure all the core in WFE */
-		online_cores &= ~(0x1 << (cpu_id * 8));
 		while (1) {
-			if (online_cores == wfe_done) {
+			if (running_cores == wfe_done) {
 				break;
 			}
 		}
diff --git a/plat/imx/imx93/ddr/dram.c b/plat/imx/imx93/ddr/dram.c
index dcd78d81b..ea49274fb 100644
--- a/plat/imx/imx93/ddr/dram.c
+++ b/plat/imx/imx93/ddr/dram.c
@@ -1,10 +1,11 @@
 /*
- * Copyright 2022, 2024 NXP
+ * Copyright 2022, 2024, 2026 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
 
 #include <common/runtime_svc.h>
+#include <lib/el3_runtime/cpu_data.h>
 #include <lib/mmio.h>
 #include <lib/spinlock.h>
 #include <drivers/delay_timer.h>
@@ -18,7 +19,7 @@
 /* lock used for DDR DVFS */
 spinlock_t dfs_lock;
 struct dram_timing_info *timing_info;
-static volatile uint32_t core_count;
+static volatile uint32_t wfe_done;
 static unsigned int num_fsp;
 static unsigned int cur_fsp;
 static volatile bool in_progress = false;
@@ -307,9 +308,12 @@ int ddrc_apply_reg_config(enum reg_type type, struct dram_cfg_param *reg_config)
 static uint64_t waiting_dvfs(uint32_t id, uint32_t flags,
 		void *handle, void *cookie)
 {
+	uint64_t mpidr = read_mpidr_el1();
+	unsigned int cpu_id = MPIDR_AFFLVL1_VAL(mpidr);
+
 	/* set the WFE done status */
 	spin_lock(&dfs_lock);
-	core_count++;
+	wfe_done |= (1 << cpu_id * 8);
 	dsb();
 	spin_unlock(&dfs_lock);
 
@@ -352,11 +356,19 @@ void dram_info_init(unsigned long dram_timing_base)
 		panic();
 }
 
+static inline aff_info_state_t dvfs_get_aff_info_state_by_idx(unsigned int idx)
+{
+	return get_cpu_data_by_index(idx,
+				     psci_svc_cpu_data.aff_info_state);
+}
+
 int dram_dvfs_handler(uint32_t smc_fid, void *handle,
 		u_register_t x1, u_register_t x2, u_register_t x3)
 {
+	uint64_t mpidr = read_mpidr_el1();
+	unsigned int cpu_id = MPIDR_AFFLVL1_VAL(mpidr);
 	unsigned int fsp_index = x1;
-	uint32_t online_cpus = x2 - 1; 
+	uint32_t running_cores = 0;
 	int ret = 0;
 
 	/* get the fsp num, return the number of supported fsp */
@@ -374,12 +386,25 @@ int dram_dvfs_handler(uint32_t smc_fid, void *handle,
 	in_progress = true;
 	dsb();
 
+	/* Check all running CPU Cores */
+	for (unsigned int i = 0; i < PLATFORM_CORE_COUNT; i++) {
+		/* Skip current CPU Core */
+		if (i == cpu_id) {
+			continue;
+		}
+		/* Perform cache maintanence ahead of reading the target CPU state */
+		flush_cpu_data_by_index(i, psci_svc_cpu_data.aff_info_state);
+		if (dvfs_get_aff_info_state_by_idx(i) == AFF_STATE_ON) {
+			running_cores |= (0x1 << (i * 8));
+		}
+	}
+
 	/* notify other core wait for scaling done */
-	trigger_intercore_dvfs_svc();
+	trigger_intercore_dvfs_svc(running_cores);
 	
 	/* Make sure all the cpu in WFE */
 	while (1) {
-		if (online_cpus == core_count)
+		if (running_cores == wfe_done)
 			break;
 	}
 	
@@ -400,7 +425,7 @@ int dram_dvfs_handler(uint32_t smc_fid, void *handle,
 		cur_fsp = fsp_index;
 
 	in_progress = false;
-	core_count = 0;
+	wfe_done = 0;
 	dsb();
 	sev();
 	isb();
-- 
2.43.0

